### DISCUSSION TITLE
Build postgrest for Raspberray Pi in CI

### DISCUSSION THREAD

**Discussion by hannesa2** [PostgREST/postgrest#3086] [Q&A] [answered]
To no mix the the question how to run in on a Raspberry Pi https://github.com/PostgREST/postgrest/discussions/3080 I open this to build it on CI for Raspberry Pi.
In short, I guess it's more useful to build it in CI, instead to let everyone build it local. 

To make it work, I guess we need 

1. a Raspberry github runner, I could provide a 24/7 Raspberry
2. an own user on Raspberry for the runner
3. a token to link it to this repository. Here I need the token from you 
![image](https://github.com/PostgREST/postgrest/assets/3314607/b63487ff-7c36-4296-9e6d-e6ddec6cb615)
4. A proper way to build. I guess the first approach would be to try it like a normal Linux build by converting it to a matrix build. Any help is welcome. 

https://github.com/PostgREST/postgrest/blob/main/.github/workflows/ci.yaml#L109-L138

Please correct me if I'm wrong.

If you like the idea and you want to share the token, I would start with 1. and 2.
The 3. must be provided from you and with 4 I can start, but I guess here we will run into issues .... here any help is welcome

**Comment by hannesa2** [comment]
On my fork (running CI on Raspberry) I found the first issue 
<img width="1108" alt="image" src="https://github.com/PostgREST/postgrest/assets/3314607/aa559bd9-8de8-4158-ae58-523bda041cf6">

This is the issue https://github.com/nixbuild/nix-quick-install-action/issues/39

**Reply by develop7** [reply]
That's because `nix-quick-install-action` [doesn't produce nix archives for ARM](https://github.com/nixbuild/nix-quick-install-action/blob/master/flake.nix#L35). Since ARM's [seem to be supported quite well](https://releases.nixos.org/?prefix=nix/nix-2.9.2/), it shouldn't be too difficult to contribute their support to `nix-q-i-action`.

**Reply by laurenceisla** [reply]
Yes, it's as @develop7 said, there could also be problems with unsupported libraries in the nix-packages. Our ARM64 builds use `cabal` instead, installed using [GHCup](https://www.haskell.org/ghcup/). You could use that as an alternative.

Check our [script to build on ARM64](https://github.com/PostgREST/postgrest/blob/main/.github/scripts/arm/build.sh), maybe it can help you.

**Reply by hannesa2** [reply]
Where do yo use this in CI ? I already investigated and I didn't found any usage

I ask, because I don't know what to specify with

https://github.com/PostgREST/postgrest/blob/main/.github/scripts/arm/build.sh#L5C1-L7C69

```
[ -z "$2" ] && { echo "Missing 2nd argument: Build environment directory name"; exit 1; }
[ -z "$3" ] && { echo "Missing 3rd argument: GHC version"; exit 1; }
```

**Reply by hannesa2** [reply]
> Check our [script to build on ARM64](https://github.com/PostgREST/postgrest/blob/main/.github/scripts/arm/build.sh), maybe it can help you.

@laurenceisla 
Btw, I found out that my Raspberry uses `ARM` and not `ARM64`. At least for the runner installation.

**Reply by laurenceisla** [reply]
Ah, TIL about GitHub self-hosted runners.

> Where do yo use this in CI ?

We use it here: 

https://github.com/PostgREST/postgrest/blob/0c65bf494d75fbb2b8650eba1385b8d2c5d042d2/.github/workflows/ci.yaml#L286

But we execute the script AS-IS, so yeah, I don't think it's useful for your workflow as a self hosted runner, since it uses ubuntu as OS. You'll still need to install libraries, etc. to build using cabal, so we'd be back to square one (your previous issue https://github.com/PostgREST/postgrest/discussions/3080). Fell free to disregard my initial suggestion.

Having said that, I'm not sure if it can be easy to solve building on armv7 using CI.

**Reply by hannesa2** [reply]
My initial goal is to read my side-project PostgreSQL database as a REST server.
To be honest, Huskel is out of my scope and I hoped for some dirty-hand-help, so I started anyhow to make it work, to have at least an initial event and I hoped for more help, compared to a link to a collection of problems.
Sorry, I guess it's the end of my journey

Really sorry that it's not working on a Raspberry

**Reply by laurenceisla** [reply]
Thanks for your initiative in taking this issue head on. We also tried to make it work for armv7 in initial iterations as you can see for the 9.0.0 build, but hit a wall and decided to concentrate our efforts only for the more demanded armv8 architecture for now.

**Comment by steve-chavez** [comment]
@hannesa2 Do you know if there's a Raspberry Pi cloud provider? Ideally we'd have it integrated on CI, so we can produce binaries on every release.

**Reply by develop7** [reply]
I've googled https://www.mythic-beasts.com/order/rpi, not sure how reliable it is though.

Another option could be https://build.opensuse.org/, a build service that has Raspbian 10 & 11 builders. Being an open source project, we're probably qualified to use their processing powers.

**Reply by hannesa2** [reply]
@steve-chavez I don't know a Raspberry Pi cloud provider, but I can offer for free that I'll serve a Raspberry pi for you 24/7. 
Please see my [initial](https://github.com/PostgREST/postgrest/discussions/3086#discussion-5921437) post.

**Comment by wolfgangwalther** [comment, marked as answer]
My goal is to eventually cross-compile the static binaries to other platforms, possibly including raspberry PI, in the future. This will not require a custom runner, but use the regular github actions CI runners. We're not there, yet, but we're tracking this in #3363 and #3281.
